### PR TITLE
Clarify goals of AbstractController

### PR DESCRIPTION
**Pull request by fabpot**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no 
| Deprecations? | yes 
| Tickets       | Refs #42418 
| License       | MIT
| Doc PR        | n/a

AbstractController should only be about HTTP-related features and we should not encourage developers to put some other logic in controllers. See #42418 for a discussion about it.

### CODE REVIEW

**Review by OskarStark — APPROVED** [review]
Make sense 👍

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 1]
```diff
@@ -50,7 +50,7 @@
 use Twig\Environment;
 
 /**
- * Provides common features needed in controllers.
+ * Provides shortcuts for HTTP related-features in controllers.
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]:
```suggestion
 * Provides shortcuts for HTTP-related features in controllers.
```

**Review by jschaedl — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 2]
```diff
@@ -370,6 +370,8 @@ protected function createFormBuilder($data = null, array $options = []): FormBui
      * Shortcut to return the Doctrine Registry service.
      *
      * @throws \LogicException If DoctrineBundle is not available
+     *
+     * @deprecated since 5.4, inject an instance of ManagerRegistry in your controller instead
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:374]: what about a runtime deprecation as well? `deprecation-contracts` is available

**Review by ro0NL — COMMENTED** [review]
i'll assume the local service refs (eg. getSubscribedServices) remain available in a protected container like till forever :)

UPGRADE file should be updated :')

**Review by derrabus — APPROVED** [review]

**Comment by wouterj** [comment]
Makes sense, although `getDoctrine()` has been there for so long that we have to update quite some docs I think. Also a friendly ping to @weaverryan as I think this will impact SymfonyCasts as well.

**Comment by zairigimad** [comment]
Makes sense, but Messenger is one of the recently added components in the AbstractControllers, I believe it's not only HTTP related feature, should this part be flagged as deprecated and moved out from the AbstractController?

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php (hunk 2 above)
Comment (fabpot) [reply src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php:374]: Done

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 3]
```diff
@@ -426,9 +430,13 @@ protected function isCsrfTokenValid(string $id, ?string $token): bool
      * Dispatches a message to the bus.
      *
      * @param object|Envelope $message The message or the message pre-wrapped in an envelope
+     *
+     * @deprecated since 5.4, inject an instance of MessageBusInterface in your controller instead
      */
     protected function dispatchMessage(object $message, array $stamps = []): Envelope
     {
+        trigger_deprecation('symfony/framework-bundle', '5.4', 'Method "%s()" is deprecated, inject an instance of MessageBusInterface in your controller instead', __METHOD__);
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]:
```suggestion
        trigger_deprecation('symfony/framework-bundle', '5.4', 'Method "%s()" is deprecated, inject an instance of MessageBusInterface in your controller instead.', __METHOD__);
```

File: src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php [hunk 4]
```diff
@@ -370,9 +370,13 @@ protected function createFormBuilder($data = null, array $options = []): FormBui
      * Shortcut to return the Doctrine Registry service.
      *
      * @throws \LogicException If DoctrineBundle is not available
+     *
+     * @deprecated since 5.4, inject an instance of ManagerRegistry in your controller instead
      */
     protected function getDoctrine(): ManagerRegistry
     {
+        trigger_deprecation('symfony/framework-bundle', '5.4', 'Method "%s()" is deprecated, inject an instance of ManagerRegistry in your controller instead', __METHOD__);
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php]:
```suggestion
        trigger_deprecation('symfony/framework-bundle', '5.4', 'Method "%s()" is deprecated, inject an instance of ManagerRegistry in your controller instead.', __METHOD__);
```

**Review by chalasr — COMMENTED** [review]
I suggest to trigger a deprecation notice from `AbstractController::get()` when the passed id is one of `message_bus`, `messenger.default_bus` or `doctrine`.

**Review by jderusse — APPROVED** [review]
👍 for cleaning this class

**Comment by fabpot** [comment]
> I suggest to trigger a deprecation notice from `AbstractController::get()` when the passed id is one of `message_bus`, `messenger.default_bus` or `doctrine`.

I will deprecate `get` and `has` in another PR as I don't think these methods make sense as they give access to a limited (and arbitrary?) set of services.

**Review by nicolas-grekas — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by fabpot** [comment]
Follow-up: #42442

**Pull request merged**